flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/CombineRecordReader.java

```diff
@@ -0,0 +1,185 @@
+import org.apache.flink.table.storage.filestore.utils.RecordIterator;
+import org.apache.flink.table.storage.filestore.utils.RecordReader;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.util.Comparator;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+
+/**
+ * Abstract combine {@link RecordReader}. Combine key-values to key-value. The input {@link
+ * RecordReader} is sorted by key and sequence number.
+ *
+ * <p>This reader needs to hold at least two record objects at the same time to perform a combined
+ * process. So the reader only accepts wrapped reader with {@link RecordIterator#supportPrevious()}.
+ */
+public abstract class CombineRecordReader implements RecordReader<KeyValue> {
+
+    private final RecordReader<KeyValue> reader;
+    private final Comparator<RowData> userKeyComparator;
+    private final RowDataSerializer userKeySerializer;
+    private final RowDataSerializer valueSerializer;
+
+    private KeyValue candidate;
+
+    public CombineRecordReader(
+            RecordReader<KeyValue> reader,
+            Comparator<RowData> userKeyComparator,
+            RowDataSerializer userKeySerializer,
+            RowDataSerializer valueSerializer) {
+        this.reader = reader;
+        this.userKeyComparator = userKeyComparator;
+        this.userKeySerializer = userKeySerializer;
+        this.valueSerializer = valueSerializer;
+    }
+
+    @Nullable
+    @Override
+    public RecordIterator<KeyValue> readBatch() throws IOException {
+        RecordIterator<KeyValue> iterator = reader.readBatch();
+        if (iterator != null) {
+            return new CombineRecordIterator(iterator);
+        } else {
+            if (candidate != null) {
+                iterator = new SingletonRecordIterator<>(candidate);
+                candidate = null;
+                return iterator;
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public void close() throws IOException {
+        this.reader.close();
+    }
+
+    /** Combine one record to another record with same key. */
+    public abstract KeyValue combine(KeyValue from, KeyValue to);
+
+    private class CombineRecordIterator implements RecordIterator<KeyValue> {
+
+        private final RecordIterator<KeyValue> iterator;
+
+        private KeyValue current;
+
+        private CombineRecordIterator(RecordIterator<KeyValue> iterator) {
+            checkArgument(iterator.supportPrevious());
+            this.iterator = iterator;
+        }
+
+        @Override
+        public boolean advanceNext() throws IOException {
+            if (candidate == null) {
+                boolean success = iterator.advanceNext();
+                if (!success) {
+                    return false;
+                }
+                candidate = iterator.current();
+            }
+
+            while (true) {
+                if (iterator.advanceNext()) {
+                    // change key, we can output candidate
```

Comment:
`change key` => `two different keys`?

Reply:
I think change key is OK. somthing like `key changed, you need do ....`